atomic_reactor/config.py

```diff
@@ -0,0 +1,568 @@
+            keys = keys.split()
+        keys = set(keys)
+        intents_matching_deprecated_keys = []
+        for entry in reversed(self.signing_intents):
+            keys_set = set(entry['keys'])
+            if (keys and keys_set >= keys) or keys == keys_set:
+                return entry
+
+            permissive_keys_set = set(entry['keys'] + entry.get('deprecated_keys', []))
+            if keys and permissive_keys_set >= keys:
+                intents_matching_deprecated_keys.append(entry)
+
+        if not intents_matching_deprecated_keys:
+            raise ValueError('unknown signing intent keys "{}"'.format(keys))
+
+        logger.warning(
+            'signing intent keys "%s" contain deprecated entries in the "%s" signing intent',
+            keys,
+            intents_matching_deprecated_keys[0]['name']
+         )
+        return intents_matching_deprecated_keys[0]
+
+
+class Configuration(object):
+    """
+    Class to parse the atomic-reactor configuration file
+    """
+    DEFAULT_CONFIG = {ReactorConfigKeys.VERSION_KEY: 1}
+
+    def __init__(self, workflow, config_path=None, basename=REACTOR_CONFIG_BASE_NAME,
+                 env_name=REACTOR_CONFIG_ENV_NAME, raw_config=None):
+
+        self.workflow = workflow
+        self.reactor_config_env = env_name
+        self.reactor_config_path = None
+        if config_path and basename:
+            self.reactor_config_path = os.path.join(config_path, basename)
+        self.conf = deepcopy(self.DEFAULT_CONFIG)
+        reactor_config_from_env = os.environ.get(self.reactor_config_env, None)
+
+        if raw_config:
+            logger.info("reading config from raw_config kwarg")
+            self.conf = deepcopy(raw_config)
+
+        elif reactor_config_from_env:
+            logger.info("reading config from %s env variable", env_name)
+            self.conf = read_yaml(reactor_config_from_env, 'schemas/config.json')
+
+        elif self.reactor_config_path and os.path.exists(self.reactor_config_path):
+            logger.info("reading config from %s", self.reactor_config_path)
+            self.conf = read_yaml_from_file_path(self.reactor_config_path, 'schemas/config.json')
+
+        else:
+            logger.info("using default config: %s", self.DEFAULT_CONFIG)
+
+        version = self.conf[ReactorConfigKeys.VERSION_KEY]
+        if version != 1:
+            raise ValueError("version %r unknown" % version)
+
+        logger.info("reading config content %s", self.conf)
+
+        self.set_workflow_based_on_config()
+
+    def set_workflow_based_on_config(self):
+        """
+        Sets attributes in workflow based on config
+        """
+
+        # set source registry and organization
+        if self.workflow.dockerfile_images:
+            source_registry_docker_uri = self.source_registry['uri'].docker_uri
+            organization = self.registries_organization
+            self.workflow.dockerfile_images.set_source_registry(source_registry_docker_uri,
```

Comment:
Does `workflow` need to be an attribute of Configuration? That seems backwards to me (configuration doesn't have a workflow, the workflow has some configuration). How about instead giving this method a `workflow` param? The `openshift_session` property would also need to be changed, probably turned into a method with a `namespace` param.

Reply:
workflow has configuration, which uses/changes workflow

also it is easy to use it like that

Reply:
Easy to use but IMO harder to understand and keep clean. But I'd like some more opinions on this

Reply:
I agree with Adam here.

Reply:
workflow removed